### PR TITLE
I27 create a contact page

### LINKED ISSUE #27: Create a Contact page
## Basic Information
The Contact should have:
- [x] A list of clients and their info
- [x] A search bar to filter clients
- [ ] UPDATE: User profile should be the first on the list
- [ ] UPDATE: Read-only tags for user profile

Contact details:
- Name: string(100)
-  Phones: string(250)
- Free text during editing - sometimes these will say things like "Speak to her daughter Anne on 0400 000 000"
[once saved, preferably then rendered with <a href="sms:..."> where [mobile number(s) can be detected](https://www.npmjs.com/package/libphonenumber-js) within the string, or <a href="tel:..."> for others]
- Email: string(250), read-only view rendered with <a href="mailto:...>
- Tags: string(250)
[Multi-select plus custom new tags](https://ej2.syncfusion.com/react/demos/#/bootstrap5/multi-select/custom-value).  These to be used for grouping / search tags similar to "Labels" in Google Contacts.
Default initial tag options: Clients, Coordinator, Volunteers.
- Address: string(250)
[nice to have Directions button for [Apple Maps](https://developer.apple.com/library/archive/featuredarticles/iPhoneURLScheme_Reference/MapLinks/MapLinks.html) for iOS devices, [Google Maps](https://gearside.com/easily-link-to-locations-and-directions-using-the-new-google-maps/#place-address) on others]
- Region: [multi-select](https://ej2.syncfusion.com/react/demos/#/bootstrap5/multi-select/default) options (& codes)
Armadale (ASJ), Busselton (BN), Central (C), Coastal South (CS), Eastern (E), Northern (N), Preston (PR), Southern (S), Western (W)
New contact Region should default to user profile region.
- Pets: string(250)
- Notes: string(max)

## Other Information
Potential scopes: 
- Add profile pic for each of the clients, need to use Firebase Storage

### PR DESCRIPTION
**Pull request by Will-H007**
## Change Summary
**[Briefly summarise the changes that you made. Just high-level stuff]**

### Change Form
**Fill this up (NA if not available). If a certain criteria is not met, can you please give a reason.**

- [x] The pull request title has an issue number
- [ ] The change works by "Smoke testing" or quick testing
- [ ] The change has tests
- [ ] The change has documentation

## Other Information
**[Is there anything in particular in the review that I should be aware of?]**

**contactinfo.tsx**
- The phone icon will call the attached number
- The mail icon opens up email 
- The marker icon opens up google maps with address
- Added an edit icon (not on figma design) which switches elements between input fields and text fields.
- Added save and cancel edit buttons, save will submit the new data, cancel will keep the original data 

**contact/[id].tsx**
- Currently uses getStaticProps/Path this will need to be changed when data is retrieved from the API

**ContactItem & ContactInfo**
- retrieves the contact image as one of the props, if we finish the MVP and continue working on it the image should be a part of the Contact type. Our code is already made to handle an image.

**ContactForm**
- When the fields update it is captured by the state contactForm.  ContactForm holds the data that will be sent to firebase.

# Related Issue

- Resolve #27

### CODE REVIEW

File: src/components/Contact/contactlist.tsx
Comment (APM246) [inline src/components/Contact/contactlist.tsx:27]: A cleaner way to do this would be using the map() function:
contactsItems = contacts.map((contact) ->(
  <ContactItem
    key={contact.id}
    userid={contact.id}
    .......
   />
)

https://reactjs.org/docs/lists-and-keys.html. It achieves the same thing but is the cleaner way of doing it

**Review by APM246 — CHANGES_REQUESTED** [review]

File: src/components/Contact/contactlist.tsx
Comment (pavo-etc) [reply src/components/Contact/contactlist.tsx]: Chantelle has just fixed this

File: src/pages/contact.tsx
Comment (APM246) [inline src/pages/contact.tsx:19]: If this code was to not change, it would be fine. However if additional state variables were added to the component, the filteredContacts variable would be re-assigned on every single render even if the searchField state wasn't changing. If the list of contacts was large, this would waste rendering time. 

I think a better way would be to use filteredContacts as a state instead of searchField. In the onSearchChange() function, the filteredContacts state could be updated.

Comment (ChantelleKerr) [reply src/pages/contact.tsx]: @APM246 I've changed it so that it uses filteredContacts as a state. Are you able to check if I did it right? 

Also off-topic but we have a lot of repeated use of the type "contactData" would it be better to put these types in its own file and import it where it's needed rather than repeatedly writing it?

Comment (APM246) [reply src/pages/contact.tsx]: Yea looks great :)

Yea putting contactData in a separate file would be good so there's only 1 consistent definition of it

Comment (ChantelleKerr) [reply src/pages/contact.tsx]: @APM246 Done, thanks :)

Do you have any suggestions on how to specify the correct type for 'context' for getStaticProps under pages/contact/[id].tsx I can only get it to work with type 'any' but the GitHub checks says not to use type 'any'. Not sure if I've done this page right.

Comment (APM246) [reply src/pages/contact.tsx]: It looks like NextJS doesn't have a declared type for the context param so you'll just have to define a custom type for it. That type declaration could go in types/types.ts (https://nextjs.org/docs/api-reference/data-fetching/get-static-props).

I think the [id].tsx looks fine, probably doesn't matter since it'll have to be changed once the API is built. Also don't forget to add the async keyword to getStaticPaths and getStaticProps

File: src/components/Contact/contactitem.tsx
Comment (NicholasChoong) [inline src/components/Contact/contactitem.tsx:13]: fragment is not needed here

File: src/components/Contact/contactlist.tsx
Comment (NicholasChoong) [inline src/components/Contact/contactlist.tsx:22]: framgnet is not needed here too

File: src/components/Contact/profileitem.tsx
Comment (NicholasChoong) [inline src/components/Contact/profileitem.tsx:14]: same thing

File: package-lock.json
Comment (NicholasChoong) [inline package-lock.json:11889]: We are using yarn so we do not need this lock file

File: package.json
Comment (NicholasChoong) [inline package.json:43]: I dont think this package is used anymore

**Review by NicholasChoong — DISMISSED** [review]

**Comment by NicholasChoong** [comment]
Looks good!

**Review by NicholasChoong — APPROVED** [review]
Great work!

File: src/components/Contact/contactform.tsx
Comment (APM246) [inline src/components/Contact/contactform.tsx:52]: If only one element with a given index needs to be modified, map() won't be needed

File: src/components/Contact/contactform.tsx
Comment (APM246) [inline src/components/Contact/contactform.tsx:94]: I think this code appears enough times that it can be made into its own component (e.g. accepting width and height props)

**Review by APM246 — CHANGES_REQUESTED** [review]
Great work, some random stuff I noticed (in addition to in-line comments):
- Cursor icon doesn't change when hovering over edit button
- On smaller screens, when a contact page has several pets/notes there is no margin on the bottom element (e.g. the "Me" contact page)
- for better clarity the pages/contact.tsx file can be moved to pages/contact/index.tsx
- Is new-tab.png used anywhere?
- don't have to do this but maybe it would be a good idea to add a //TODO anywhere CONTACT_DATA is used

**Comment by ChantelleKerr** [comment]
> * Cursor icon doesn't change when hovering over edit button 

DONE

> * On smaller screens, when a contact page has several pets/notes there is no margin on the bottom element (e.g. the "Me" contact page)

I can't seem to replicate this problem. For me, there is a margin at the bottom. Are you able to please clarify what you mean? 
> * Is new-tab.png used anywhere?

DONE - I removed it because it's not used
> * don't have to do this but maybe it would be a good idea to add a //TODO anywhere CONTACT_DATA is used

DONE

Thanks for reviewing I'll get to the other requested changes later tonight :)

File: src/pages/contact.tsx
Comment (pavo-etc) [reply src/pages/contact.tsx]: > Lines 31-39 and Lines 47-56 look the same so they could be abstracted to a function

This has now been fixed

Comment (wambampham) [reply src/pages/contact.tsx]: Fixed

**Comment by pavo-etc** [comment]
All requested changes have been fixed

**Review by APM246 — DISMISSED** [review]
Should be good to merge once merge conflicts are resolved :)

**Review by NicholasChoong — APPROVED** [review]
Looks good!

**Review by skyheat — APPROVED** [review]
Looks good!

**Pull request merged**
